Add more detail around our testing steps

tahb opened the pull request:
# Changes in this PR

Based upon https://github.com/ministryofjustice/hmpps-temporary-accommodation-ui/pull/196 so please help to merge that first.

This change details the testing steps to help explain the differences of the numerous options.

As I said in the commit, the `test:e2e` commands don't work for me and I'm not sure what their original intent was, so I've left them out.

# Release checklist

As part of our continuous deployment strategy we must ensure that this work is
ready to be released at any point. Before merging to `main` we must first
confirm:

## Pre merge checklist

- [ ] Have all changes to any dependencies been deployed to both `preprod` and
    `production` in a backwards compatible way? (This includes our API,
    infrastructure, third-party integrations and any secrets or environment variables)
- [ ] Has a required data migration been included in this change or been done in
    advance?

## Post merge checklist

Once we've merged we now need to release this through to production before
considering it done.

[Find the build-and-deploy job in CircleCI](https://app.circleci.com/pipelines/github/ministryofjustice/hmpps-temporary-accommodation-ui)
and work through the following steps:

- [ ] Has the product manager asked to provide approval for this feature?
  - [ ] Has the product manager approved?
- [ ] Manually approve release to preprod
- [ ] Verify change released to preprod
- [ ] Manually approve release to prod
- [ ] Verify change released to production

Should a release fail at any step, you as the author should now lead the work to
fix it as soon as possible.

tahb committed:
Both integration tests require wiremock

For me running `npm run test:integration` blew up with various errors including:

```
From Node.js Internals:
      at <unknown> (      Error: connect ECONNREFUSED ::1:9999)
      at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1494:16)
```

This (via Docker Compose) points to the absence of a wiremock instance.

If we take a look at `npm run test:integration:ui` this does start wiremock. This makes sense as we expect our integration tests to be mocked out. If after running this you go back to `npm run test:integration` it will magically start working again! This is due to the fact the wiremock container wasn't removed and so allows the tests to work accidentally.

tahb committed:
Describe common ways to test

Whilst script/test takes care of all the testing steps for us it isn't often practical to run the whole thing each time as it takes a long time.

During dev we may want to run individual tests that are faster to get feedback from.

I couldn't figure out exactly what the purpose of the `npm run test:e2e` was as it doesn't seem to work currently, so I've left that out for now.

bryangaledxw commented on package.json:
```diff
@@ -23,7 +23,7 @@
     "test:e2e": "cypress run -C cypress.config.e2e.ts",
     "test:e2e:ui": "set -a && source e2e.env && set +a && cypress open --e2e --browser electron -C cypress.config.e2e.ts --config baseUrl=https://approved-premises-dev.hmpps.service.justice.gov.uk",
-    "test:integration": "start-server-and-test start-feature http://localhost:3007/ping int-test",
+    "test:integration": "npm run start-test-wiremock && start-server-and-test start-feature http://localhost:3007/ping int-test",
```
I nearly did this tidy up myself, then backed out given that `test:integration` is run in `script/test`, and I wasn't sure what would happen if that script spun up wiremock via `docker-compose`, spun wiremock up again via `test:integration`, then only cleaned up the latter. Happy that this is safe?

tahb replied:
I think it's ok? It looks to me like both `test:integration` and `script/test` attempt to start the backing services via docker-compose but only `script/test` cleans up after itself. 

If `script/test` runs after a lone request to do `test:integration` then it should find existing docker containers and continue as normal. I've given this a test and it seems to be happy!

bryangaledxw reviewed (commented):
I feel like this might be best practice that I'm not aware of, and this isn't a strongly held objection, but I'm not sure of the need to keep the E2E environment setup out of git? I feel like if it needs to contain sensitive data, that's an issue in itself?

Is this correct? - "I found that the .e2e.env file is git ignored but also required by the local_e2e script" - I don't have `.e2e.env` anywhere and it looks like `local_e2e` only references `e2e.local.env` right now

And yep, I never use `npm run test:e2e`. I think it's assuming everything is already up and running via ap tools?

tahb committed:
Remove unused e2e testing commands

This arrived with the fork from the AP frontend. In practice we haven't had the need to test locally against a development API. The local API works for us through the use of script/local_e2e script.

tahb commented:
As we're happy to drop the following tasks - which were the remaining reference to e2e.env - I think we can go back to e2e.local:

```
"test:e2e"
"test:e2e:ui"
```

I agree that e2e.env should include safe values that we can commit to git. Looks like we are inheriting the decision to include that file in the gitignore. Feels risky to unignore that file now just incase anyone has a file they might then commit.

bryangaledxw reviewed (approved):
LGTM
